src/ripple/basics/promises.h

```diff
@@ -0,0 +1,958 @@
+        // passing to them `lhs`.
+
+        // Reap the callbacks from `rhs`.
+        decltype(storage_.callbacks_) callbacks(
+            std::move(rhs->storage_.callbacks_));
+        std::destroy_at(&rhs->storage_.callbacks_);
+        // Make `rhs` link to `lhs`.
+        std::construct_at(&rhs->storage_.link_, lhs->shared_from_this());
+        rhs->status_.store(LINKED, std::memory_order_release);
+
+        if (lprev == PENDING || lprev == LOCKED)
+        {
+            std::move(
+                callbacks.begin(),
+                callbacks.end(),
+                std::back_inserter(lhs->storage_.callbacks_));
+            lhs->status_.store(lprev, std::memory_order_release);
+        }
+        else if (lprev != CANCELLED)
+        {
+            for (auto& cb : callbacks)
+            {
+                lhs->scheduler_.schedule(
+                    [self = lhs->shared_from_this(), cb = std::move(cb)]() {
+                        cb(std::move(self));
+                    });
+            }
+        }
+
+        return true;
+    }
+
+    void
+    subscribe(callback_type&& cb)
+    {
+        auto self = this;
+        State previous = transition_(self, WRITING);
+        if (previous != PENDING && previous != LOCKED)
+        {
+            // The promise is settled. No longer taking subscribers.
+            if (previous != CANCELLED)
+            {
+                self->scheduler_.schedule(
+                    [self = self->shared_from_this(), cb = std::move(cb)]() {
+                        cb(std::move(self));
+                    });
+            }
+            return;
+        }
+        self->storage_.callbacks_.push_back(std::move(cb));
+        self->status_.store(previous, std::memory_order_release);
+    }
+
+    template <typename F>
+    decltype(auto)
+    then(F&& f)
+    {
+        using R = std::invoke_result_t<F, pointer_type const&>;
+        using Fulfiller = detail::Fulfiller<R>;
+        auto q = scheduler_.pending<typename Fulfiller::value_type>();
+        auto cb = [q, f = std::move(f)](pointer_type const& p) mutable {
+            try
+            {
+                Fulfiller{q}.fulfillWith(std::move(f), std::move(p));
+            }
+            catch (...)
+            {
+                q->reject(std::current_exception());
+            }
+        };
+        subscribe(cb);
+        return q;
+    }
+
+    // TODO: Any way to distinguish calls to this method
+    // if it were an overload of `then`?
+    template <typename F>
+    decltype(auto)
+    thenv(F&& f)
+    {
+        return then([f = std::forward<F>(f)](auto const& p) {
+            // Callbacks should never be called with a linked promise.
+            auto state = p->state_();
+            if (state == REJECTED)
+            {
+                std::rethrow_exception(p->error());
+            }
+            assert(state == FULFILLED);
+            return f(p->value());
+        });
+    }
+
+    decltype(auto)
+    reify() const
```

Comment:
is there any specific reason why you chose this name for the function? I feel `resolve_promise` (something to that effect) is easier to understand

Reply:
I think [reification](https://en.wikipedia.org/wiki/Reification_(computer_science)) is a well-known term in computer science. I do not like "resolve" here because it is too easily confused with "fulfill" and suggests transitioning the promise from an idle state to a terminal state, which is not what this method does. This method takes whatever value or exception existed in the fulfilling context and reconstitutes it in the current context, by either returning a value or throwing an exception. I'm open to calling it `reconstitute` or `deliver` or something, but not `resolve`.